File: ai-engine/app/routers/bids.py
from fastapi import APIRouter, Header, HTTPException
from pydantic import BaseModel

from app.core.config import settings
from app.services import bid_scanner

router = APIRouter(prefix="/bids", tags=["bids"])


class ScanRequest(BaseModel):
    keywords: list[str]
    company_profile: str = ""
    eligibility_threshold: float = 0.6
    start_date: str | None = None
    end_date: str | None = None


class BidResult(BaseModel):
    external_bid_no: str | None = None
    title: str | None = None
    agency: str | None = None
    matched_keyword: str | None = None
    announce_date: str | None = None
    deadline: str | None = None
    url: str | None = None
    eligibility_score: float | None = None
    ai_summary: str | None = None
    ai_judgement: str | None = None
    eligible: bool = False


class ScanResponse(BaseModel):
    results: list[BidResult]
    fetched: int
    range_start: str
    range_end: str
    judged: int
    unjudged: int


def verify_api_key(x_api_key: str) -> None:
    if x_api_key != settings.ai_engine_api_key:
        raise HTTPException(status_code=401, detail="유효하지 않은 API 키입니다.")


@router.post("/scan", response_model=ScanResponse)
def scan(payload: ScanRequest, x_api_key: str = Header(default="")) -> ScanResponse:
    verify_api_key(x_api_key)

    outcome = bid_scanner.scan(
        payload.keywords,
        payload.company_profile,
        payload.eligibility_threshold,
        payload.start_date,
        payload.end_date,
    )
    return ScanResponse(
        results=[BidResult(**item) for item in outcome["results"]],
        fetched=outcome["fetched"],
        range_start=outcome["range_start"],
        range_end=outcome["range_end"],
        judged=outcome["judged"],
        unjudged=outcome["unjudged"],
    )

File: frontend/src/api/bids.ts
import { apiClient, type ApiResponse } from './client'

export const MAX_KEYWORDS = 15

export interface BidKeyword {
  id: number
  keyword: string
  active: boolean
}

export interface BidNotice {
  id: number
  title: string
  agency: string | null
  matchedKeyword: string | null
  announceDate: string | null
  deadline: string | null
  url: string | null
  eligibilityScore: number | null
  aiSummary: string | null
  aiJudgement: string | null
  status: 'ELIGIBLE' | 'INELIGIBLE'
  crawledAt: string
}

export interface BidScanSummary {
  fetched: number
  matched: number
  newNotices: number
  eligibleCount: number
  judged: number
  unjudged: number
  rangeStart: string
  rangeEnd: string
}

export async function listKeywords(): Promise<BidKeyword[]> {
  const { data } = await apiClient.get<ApiResponse<BidKeyword[]>>('/api/bids/keywords')
  return data.data
}

export async function addKeyword(keyword: string): Promise<BidKeyword> {
  const { data } = await apiClient.post<ApiResponse<BidKeyword>>('/api/bids/keywords', { keyword })
  return data.data
}

export async function removeKeyword(id: number): Promise<void> {
  await apiClient.delete(`/api/bids/keywords/${id}`)
}

/** 조회기간 내 적격 공고 (최대 50건, 최신순). 둘 다 비우면 서버 기본값(최근 7일) 사용 */
export async function fetchRecentEligible(startDate?: string, endDate?: string): Promise<BidNotice[]> {
  const params = startDate && endDate ? { startDate, endDate } : {}
  const { data } = await apiClient.get<ApiResponse<BidNotice[]>>('/api/bids/eligible', { params })
  return data.data
}

export async function fetchBidDetail(id: number): Promise<BidNotice> {
  const { data } = await apiClient.get<ApiResponse<BidNotice>>(`/api/bids/${id}`)
  return data.data
}

/** startDate/endDate 둘 다 비우면 서버 기본값(최근 7일)을 사용 */
export async function scanNow(startDate?: string, endDate?: string): Promise<BidScanSummary> {
  const body = startDate && endDate ? { startDate, endDate } : {}
  const { data } = await apiClient.post<ApiResponse<BidScanSummary>>('/api/bids/scan-now', body)
  return data.data
}
